### DISCUSSION TITLE
Two Factor Recovery Codes

### DISCUSSION THREAD

**Discussion by AaronSadlerUK** [umbraco/Umbraco-CMS#12892] [Features and ideas]
This isn't touched on in the two factor documentation, how do we go about generating the recovery codes?

In dotnet identity we should be able to do the following:

`private readonly UserManager<MemberIdentityUser> _userManager;`

```
var codes = await _userManager.GenerateNewTwoFactorRecoveryCodesAsync(member, 6);
TempData["RecoveryCodes"] = codes;
```

This would allow them to be shown one time on the confirmation page.

However for the life of me, I cannot figure out how to get Umbraco to do this with members.

**Comment by dave0504** [comment]
Hi Arron,

maybe this way???

inject `private readonly IMemberManager _memberManager;`

Then 
`var member = await _memberManager.FindByEmailAsync(email);`
`var token= await _memberManager.GenerateTwoFactorTokenAsync(member);`

**Reply by AaronSadlerUK** [reply]
That doesn't seem to be the same thing I'm trying to use.

The recovery codes are one time codes which can be used insyead of the authenticator app, for example when it is lost.

According to Microsoft it's in UserManager, however I cannot get it to generate any codes
